Frameworks/CoreFoundation/NumberDate.subproj/CFTimeZone.c

```diff
@@ -1331,106 +1085,96 @@ CFTimeZoneRef CFTimeZoneCreateWithName(CFAllocatorRef allocator, CFStringRef nam
-        if (-14 <= hours && hours <= 14) {
-            CFTimeInterval ti = hours * 3600.0;
-            if (6 == len) {
-            return CFTimeZoneCreateWithTimeIntervalFromGMT(allocator, ('-' == buffer[3] ? -1.0 : 1.0) * ti);
-            } else {
-            if (('0' <= buffer[6] && buffer[6] <= '9') && ('0' <= buffer[7] && buffer[7] <= '9')) {
-                int32_t minutes = (buffer[6] - '0') * 10 + (buffer[7] - '0');
-                if ((-14 == hours && 0 == minutes) || (14 == hours && 0 == minutes) || (0 <= minutes && minutes <= 59)) {
-                ti = ti + minutes * 60.0;
-                return CFTimeZoneCreateWithTimeIntervalFromGMT(allocator, ('-' == buffer[3] ? -1.0 : 1.0) * ti);
+    // Check for GMT timezone names
+    // GMT timezone names must follow a specific format
+    if (6 == len || 8 == len || 9 == len) {
+        UniChar buffer[9];
+        CFStringGetCharacters(name, CFRangeMake(0, len), buffer);
+        if ('G' == buffer[0] && 'M' == buffer[1] && 'T' == buffer[2] && ('+' == buffer[3] || '-' == buffer[3])) {
+            // GMT+... or GMT-...
+            if ('0' <= buffer[4] && buffer[4] <= '9') {
+                if ('0' <= buffer[5] && buffer[5] <= '9') {
+                    // GMT+##... or GMT-##...
+                    int32_t hours = (buffer[4] - '0') * 10 + (buffer[5] - '0');
+                    if (-14 <= hours && hours <= 14) {
+                        CFTimeInterval ti = hours * 3600.0;
+                        if (6 == len) {
+                            // GMT+## or GMT-##
+                            return CFTimeZoneCreateWithTimeIntervalFromGMT(allocator, ('-' == buffer[3] ? -1.0 : 1.0) * ti);
+                        } else if ((8 == len) && ('0' <= buffer[6] && buffer[6] <= '9') && ('0' <= buffer[7] && buffer[7] <= '9')) {
+                            // GMT+#### or GMT-####
+                            int32_t minutes = (buffer[6] - '0') * 10 + (buffer[7] - '0');
+                            if ((-14 == hours && 0 == minutes) || (14 == hours && 0 == minutes) || (0 <= minutes && minutes <= 59)) {
+                                ti = ti + minutes * 60.0;
+                                return CFTimeZoneCreateWithTimeIntervalFromGMT(allocator, ('-' == buffer[3] ? -1.0 : 1.0) * ti);
+                            }
+                        } else if ((9 == len) && (buffer[6] == ':') && ('0' <= buffer[7] && buffer[7] <= '9') && ('0' <= buffer[8] && buffer[8] <= '9')) {
+                            // GMT+##:## or GMT-##:##
+                            int32_t minutes = (buffer[7] - '0') * 10 + (buffer[8] - '0');
+                            if ((-14 == hours && 0 == minutes) || (14 == hours && 0 == minutes) || (0 <= minutes && minutes <= 59)) {
+                                ti = ti + minutes * 60.0;
+                                return CFTimeZoneCreateWithTimeIntervalFromGMT(allocator, ('-' == buffer[3] ? -1.0 : 1.0) * ti);
```

Comment:
it seems there are too many things going on here.Given we are looking for specific things. it's easy to make mistakes here (now+future maintenance).
This seems like it can be done with using regex.